service/paxos/paxos_state.cc

```diff
@@ -219,19 +219,29 @@ future<> paxos_state::learn(storage_proxy& sp, db::system_keyspace& sys_ks, sche
                     on_internal_error(logger, format("schema version in learn does not match current schema"));
                 }
 
-                return sp.mutate_locally(schema, decision.update, tr_state, db::commitlog::force_sync::yes, timeout);
+                return sp.mutate_locally(schema, decision.update, tr_state, db::commitlog::force_sync::yes, timeout).then([&sys_ks, &decision, schema, timeout] (std::optional<db::view::update_backlog> backlog) {
+                    // We don't need to lock the partition key if there is no gap between loading paxos
+                    // state and saving it, and here we're just blindly updating.
+                    return utils::get_local_injector().inject("paxos_timeout_after_save_decision", timeout, [&sys_ks, &decision, schema, timeout, backlog] {
+                        return sys_ks.save_paxos_decision(*schema, decision, timeout).then([backlog] {
+                            return make_ready_future<std::optional<db::view::update_backlog>>(std::move(backlog));
+                        });
```

Comment:
I duplicated the code here because, as a result of this change, the `f = f.then(...)` now changes the type of `f`. I used another approach in the rebase